### ISSUE TITLE
Add more OID's to X509 signature algorithm

### ISSUE THREAD

**Issue by dol** [phpseclib/phpseclib#770]
The x509 parser is missing some OID for the mapping of the signature algorithm.

A list of OID's can be found in the following project:
https://golang.org/src/crypto/x509/x509.go (Line 232)
https://cryptography.io/en/latest/x509/reference/#cryptography.x509.oid.SignatureAlgorithmOID

An other problem is, that OID are replaced with the NOID mapping. If no NOID was found, the signature algorithm will contain the OID. This makes it very hard to identify the used algorithm.
It would be nice to have separation of the NOID and the OID.

**Comment by terrafrost** [comment]
I can add more OID's.

As for NOID...  I have not heard of that. What is that?

The ASN.1 description for AlgorithmIdentifier is as follows:

```
   AlgorithmIdentifier  ::=  SEQUENCE  {
        algorithm               OBJECT IDENTIFIER,
        parameters              ANY DEFINED BY algorithm OPTIONAL  }
```

I guess OBJECT IDENTIFIER could correspond to an OID of that just happened to have a mapping of  "noid" or something and then parameters could be more dynamic or something but, in any event, I'd need more information.

If there is an OID that maps to "noid" in what RFC is that mentioned in?

**Comment by dol** [comment]
Sorry, NOID stands for named OID or ASN.1 notation. It's the translation of OID to it's notation name. This name is IMHO not defined in any RFC.
In your [code](https://github.com/phpseclib/phpseclib/blob/d9e9504fbacc20502e4b456891b6f310e8022e91/phpseclib/File/X509.php#L1321) a mapping is performed from a OID to it's name.

This is a pseudo code to illustrate the problem:
```php
<?php
include('File/X509.php');

$x509 = new File_X509();
$cert = $x509->loadX509('...');

$algo = $cert['signatureAlgorithm']['algorithm'];
// If the OID was found in the mapping a string is returned. The original OID information will be overwritten.
// '1.2.840.113549.1.1.5'  => 'sha1WithRSAEncryption'
// '1.2.840.10045.4.3.3' => stays '1.2.840.10045.4.3.3'

switch ($algo) {
    case 'sha1WithRSAEncryption':
        echo 'SHA1 with RSA';
        break;
    // This will be replaced with 'sha384WithECDSAEncryption'
    case '1.2.840.10045.4.3.3':
        echo 'SHA384 with ECDSA';
        break;
}

// This code will break, it phpseclib later adds a mapping for '1.2.840.10045.4.3.3'.
// In the future there will always be OID that don't have a corresponding label/notation.
// And people have old version of phpseclib
```

**Comment by dol** [comment]
Additional note:
The PHP core (since 5.5.24, 5.6.7) returns 3 values for the signature algorithm when using openssl_x509_parse:
- signatureTypeSN -> openssl short name
- signatureTypeLN -> openssl long name
- signatureTypeNID -> OID

Corresponding bug report: https://bugs.php.net/bug.php?id=67403

**Comment by terrafrost** [comment]
I guess I could make it so an object is returned for OIDs and you could do...

`$oid->getOID();` returns OID
`$oid->getName();` returns name or false
`$oid->__toString();` returns name or OID

But that would be a BC breaking change and would only go into the master branch.

Another possibility:

`$x509->setOIDMode(FILE_X509_NAME_OID);` would be the current (default) behavior
`$x509->setOIDMode(FILE_X509_OID);` would make it return OIDs only

This possibility would add new functionality but it wouldn't change existing functionality (unless you explicitly changed it yourself) so it wouldn't be a BC break and could go even in the 1.0 branch.

**Comment by dol** [comment]
I would perform the following steps:
1. Add $x509->setOIDMode(); to 1.0 and 2.0 minor version
2. Introduce BC break in v3.0

An other idea is to move x509 to it's own library for handling certificates. But this needs a lot of work to split phpseclib into smaller components. A possible proposal would look like this:
- Repository of phpseclib\File\ASN1
- Repository that extracts informations from ASN1 objects
  - Read RSA/EC/DSA private/public in the format of PKCS#1/5/8
  - Read X.509 ceritificate, CRL and CSR(PKCS#10)
- Repository of Crypto implementations
  - phpseclib\Crypt part -> reads key informations from ASN.1 parsing
  - phpseclib\Net part

This would solve IMHO some problems that I see. First of all it's nice to have a solid pure PHP implementation of cryptographic primitives. Due to the lack of a proper ECDSA support of the openssl extension in PHP I used the [ASN.1 parser](https://github.com/namshi/jose/blob/bc8fa0d743932424b172dcb4b88aacb5788ed367/src/Namshi/JOSE/Signer/OpenSSL/ECDSA.php#L40). To stop reinvent the wheel it would be nice if other people could profit from this work.

**Comment by terrafrost** [comment]
> An other idea is to move x509 to it's own library for handling certificates.

That actually is the plan. Well, creating File_CSR, File_CRL, File_SPKAC, etc. Putting them in their own repo's, otoh...  well maybe it'll happen someday but I have no timetable for that.

Anyway, I'll try to do this within the next few days. This week wound up being pretty busy for me so I didn't get a chance to do it earlier.

**Comment by terrafrost** [comment]
#789 addresses this. I changed how I did it. Trying to do `$x509->setOIDMode(FILE_X509_NAME_OID);` would have just been too problematic. Let's say you set the OID mode and then load the X509 cert. X509.php is looking for OID's by name - not by OID. I guess you could reload the X509 cert - reparse it - depending on the mode but that just seems like overkill.

I think the getOID() method works just fine.

In the future I'll probably make it static but there are a couple other methods I want to make static as well. I'll just do that when I do those.

**Comment by dol** [comment]
Ok. Thanks. Is it possible to patch also apply this fix also to the 1.0 version.

**Comment by bantu** [comment]
@dol #789 is against 1.0

**Comment by dol** [comment]
@bantu Sorry. My bad.

**Comment by terrafrost** [comment]
Fixed by #789
